internal_worker_base.go

```diff
@@ -141,33 +150,42 @@ func (bw *baseWorker) Stop() {
 	// TODO: The poll is longer than the 10 seconds, we probably need some way to hard terminate the
 	// poll routines as well.
 
-	if success := awaitWaitGroup(&bw.shutdownWG, 10*time.Second); !success {
+	if success := awaitWaitGroup(&bw.shutdownWG, 2*time.Second); !success {
 		bw.logger.Info("Worker timed out on waiting for shutdown.")
 	}
 }
 
+func (bw *baseWorker) Done() <-chan os.Signal {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	return c
+}
+
 // execute handler wraps call to process a task.
 func (bw *baseWorker) execute(routineID int) {
 	for {
-		// Check if we have to backoff.
-		// TODO: Check if this is needed concurrent retires (or) per connection retrier.
-		bw.retrier.Throttle()
-
-		// Check if we are rate limited
-		if !bw.rateLimiter.Consume(1, time.Millisecond) {
-			continue
-		}
-
-		err := bw.options.taskPoller.PollAndProcessSingleTask()
-		if err == nil {
-			bw.retrier.Succeeded()
-		} else if isClientSideError(err) {
-			bw.logger.Info("Poll and processing failed with client side error", zap.Int(tagRoutineID, routineID), zap.Error(err))
-			// This doesn't count against server failures.
-		} else {
-			bw.logger.Info("Poll and processing task failed with error", zap.Int(tagRoutineID, routineID), zap.Error(err))
-			bw.retrier.Failed()
-		}
+		ch := make(chan struct{})
+		go func() {
```

Comment:
tchannel.Close lives here:
https://github.com/uber/tchannel-go/blob/dev/channel.go#L747

but its not a trivial change, because the channel gets created at the app, not the library today, see example here:
https://github.com/samarabbas/cadence-samples/blob/master/cmd/samples/common/factory.go#L96

So, yes, we should tackle this separately. Let's merge this PR / fix